crypto/edhoc-crypto-cryptocell310-sys/src/lib.rs

```diff
@@ -18,183 +20,273 @@ fn convert_array(input: &[u32]) -> [u8; SHA256_DIGEST_LEN] {
+        let mut crys_public_key: *mut CRYS_ECPKI_UserPublKey_t =
+            &mut CRYS_ECPKI_UserPublKey_t::default();
+        let mut temp_data: *mut CRYS_ECPKI_KG_TempData_t = &mut CRYS_ECPKI_KG_TempData_t::default();
+        let mut temp_fips_buffer: *mut CRYS_ECPKI_KG_FipsContext_t =
+            &mut CRYS_ECPKI_KG_FipsContext_t::default();
+
+        unsafe {
+            CRYS_ECPKI_GenKeyPair(
+                &mut rnd_context as *mut _ as *mut c_void,
+                rnd_generate_vect_func,
+                curve_256,
+                crys_private_key,
+                crys_public_key,
+                temp_data,
+                temp_fips_buffer,
+            );
+        }
 
-    let mut public_key_compressed = [0x0u8; P256_ELEM_LEN + 1];
-    public_key_compressed[0] = 0x02;
-    public_key_compressed[1..].copy_from_slice(&public_key[..]);
+        let mut private_key: [u8; P256_ELEM_LEN] = [0x0; P256_ELEM_LEN];
+        let mut key_size: u32 = P256_ELEM_LEN.try_into().unwrap();
 
-    let mut public_key_cc310: CRYS_ECPKI_UserPublKey_t = Default::default();
+        unsafe {
+            CRYS_ECPKI_ExportPrivKey(crys_private_key, private_key.as_mut_ptr(), &mut key_size);
+        }
 
-    let mut domain =
-        unsafe { CRYS_ECPKI_GetEcDomain(CRYS_ECPKI_DomainID_t_CRYS_ECPKI_DomainID_secp256r1) };
+        // let private_key = BytesP256ElemLen::from_public_slice(&private_key[..]);
 
-    unsafe {
-        _DX_ECPKI_BuildPublKey(
-            domain,
-            public_key_compressed.as_mut_ptr(),
-            (P256_ELEM_LEN + 1) as u32,
-            EC_PublKeyCheckMode_t_CheckPointersAndSizesOnly,
-            &mut public_key_cc310,
-            core::ptr::null_mut(),
-        );
-    }
+        let mut public_key: [u8; P256_ELEM_LEN + 1] = [0x0; P256_ELEM_LEN + 1];
+        let mut key_size: u32 = (P256_ELEM_LEN as u32) + 1;
+        let compressed_flag: CRYS_ECPKI_PointCompression_t =
+            CRYS_ECPKI_PointCompression_t_CRYS_EC_PointCompressed;
 
-    let mut private_key_cc310: CRYS_ECPKI_UserPrivKey_t = Default::default();
+        unsafe {
+            CRYS_ECPKI_ExportPublKey(
+                crys_public_key,
+                compressed_flag,
+                public_key.as_mut_ptr(),
+                &mut key_size,
+            );
+        }
 
-    unsafe {
-        CRYS_ECPKI_BuildPrivKey(
-            domain,
-            private_key.clone().as_mut_ptr(),
-            P256_ELEM_LEN as u32,
-            &mut private_key_cc310,
-        );
-    }
+        let public_key: [u8; P256_ELEM_LEN] = public_key[1..33].try_into().unwrap(); // discard sign byte
 
-    unsafe {
-        CRYS_ECDH_SVDP_DH(
-            &mut public_key_cc310,
-            &mut private_key_cc310,
-            output.as_mut_ptr(),
-            &mut output_len,
-            &mut tmp,
-        );
+        (private_key, public_key)
     }
-
-    output
 }
 
 fn hmac_sha256(message: &mut [u8], mut key: [u8; SHA256_DIGEST_LEN]) -> BytesHashLen {
```

Comment:
This could be within an `impl Crypto` just like in the `psa` case, just for consistency.

Reply:
There is an actual difference -- the `psa` one makes use of self, whereas this hmac_sha256 implementation doesn't use a self. Then again, that's just because we have that `unsafe`, so probably it is indeed a good thing, anticipating the things that can now be reaped in terms of doing stuff at startup time.